### DISCUSSION TITLE
Make it easier to support a custom compiler

### DISCUSSION THREAD

**Discussion by kobalicek** [microsoft/vcpkg#37544] [Q&A]
Using vcpkg is pretty simple unless you need to do something like using a non-default compiler. For example if I have a cmake project, I can just set `CC` and `CXX` environment variables and when running CMake it will use the compiler as specified by the env vars. This is a common practice in Linux world and most tools would respect these variables.

Now... If you have a project with a dependency, and you want to test multiple compilers on your machine, it becomes a nightmare of scripting and trials / errors. The biggest problem is that vcpkg requires you to write custom triplets and cmake toolchains, which requires a lot of scripting and it's frustrating when you hit `vcpkg install` and you get dependencies built with a different C++ compiler, often manifesting at link time.

So, I have decided to report this as an issue.

I think that vcpkg should AT LEAST respect the environment variables set. Because I really think that this is a problem. I setup `CC` and `CXX` (possibly other env vars as well), run `vcpkg install`, and it installs dependencies compiled with a different compiler than cmake would pick when actually configuring the main project.

For example on macos the default "Apple Clang" compiler doesn't come with openmp, so packaes that actually use openmp are all broken by design, which means that I cannot install "faiss", for example, because the default compiler cannot compile it.

Is there going to be a "simple" solution to this problem or are we really required to write tons of scripts and toolchains just to get what two environment variables can do? Is there any recommendation of how to do this in a simple and straightforward way? I have two environment variables `CC` and `CXX`, and I want vcpkg to use them.

**Comment by kobalicek** [comment]
I would just add simple scenarios that I would expect `vcpkg` to handle like all other tools:

  1. I have a project that has a dependency. In the past it used git submodule, now I wanted a modern approach so I started using `vcpkg`. The user wants to compile it with clang-18 on Linux. In the past, with git submodule, the only thing needed to do that was to properly set `CC` and `CXX` variables, or to use `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` definitions during configure time. Move to `vcpkg` frustrates users as they cannot specify the compiler anymore in a simple way. There is no option to set a compiler...

  2. Continuous integration. Build matrix has 100 combinations having platforms, compilers, etc... Again, setting up the compiler with vcpkg is a nightmare as you have to mess with triplets and toolchains, something that is totally not required when using just cmake.

I'm really frustrated about this, because I don't want to write scripts that generate vcpkg triplets to just add a "chained" toolchain file into these triplets... and even if I manage to actually generate these files, the chained toolchain files are missing stuff, because the only thing I want to just override the compiler and let cmake handle the rest. So to be honest, I don't want to generate triplets and toolchains, why should that be necessary if the only thing I want to do is to change a compiler?

I consider this broken, because this is totally non-transparent to users that expect `vcpkg` followed by `cmake` to just do the right thing.

**Comment by kobalicek** [comment]
BTW one additional thing is that:

   `VCPKG_KEEP_ENV_VARS=CC;CXX`

doesn't work either. Even if I tell `vcpkg` explicitly I want to keep these environment variables they would not be used by the build...

**Comment by dg0yt** [comment]
Other than presented, `CC` and `CXX` are used at least for some triplets if not using Windows. (The environment isn't cleaned, and CMake picks these variables.) 
~~~
$ CC=false CXX=false vcpkg install rhash
~~~
fails quickly, logging
~~~
-- The C compiler identification is unknown
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: /bin/false
-- Check for working C compiler: /bin/false - broken
-- Configuring incomplete, errors occurred!
~~~

However, I wouldn't expect this to work well e.g. for cross builds. Nobody tells cmake when to prefer `CC_FOR_BUILD` instead.
And there are ports and scripts which do not properly forward the detected CMake toolchain settings to alternative build systems.

**Comment by kobalicek** [comment]
This doesn't work for me on macos - it always uses Apple Clang and not Homebrew clang as I have specified in `CC` and `CXX`.

**Reply by dg0yt** [reply]
But `scripts/toolchains/osx.cmake` and `scripts/buildsystems/vcpkg.cmake` don't set `CMAKE_<LANG>_COMPILER` at all.
`CC=false CXX=false vcpkg install rhash` has the expected effect on my MBA, too.
Which port are you trying to build?

**Reply by kobalicek** [reply]
I'm trying to build faiss on macos. I can build it without vcpkg on my own, but with vcpkg it's a nightmare, and I want to build it because it can build other 10 deps just okay.

I will review my code, maybe I'm doing something wrong, because it's a project that needs some additional scripting to prepare the environment, etc..
